### PR TITLE
enip: convert to rust

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3958

Describe changes:
- convert enip parser to rust

Alon the way, also
- enip_command keyword accepts now string enumeration as values.
- transactions are now bidirectional
- there is a enip logger
- gap support is improved with probing for resync
- SEQUENCE_ADDR_ITEM value is fixed to 0x8002 instead of 0xB002

Draft as is this is not complete :
- need to fix S-V number of alerts for enip-keywords (off by one)
- add more S-V test for enip_command with string , for logger, complete json schema
- need better response/request association for bidirectional transactions
- get cip attributes by parsing deeper based on cip service, and reuse that for `cip_service` keyword
- parse more, log more, add more keywords (enip.status)...
- set event on parsing error
- use frames ?
- take a look into https://redmine.openinfosecfoundation.org/issues/6304

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 16594

File: rust/src/enip/detect.rs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+            let boxed = Box::new(ctx);
+            return Box::into_raw(boxed) as *mut _;
+        }
+    }
+    return std::ptr::null_mut();
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_cip_service_free(ctx: *mut c_void) {
+    // Just unbox...
```
Comment (jasonish) [inline rust/src/enip/detect.rs:175]: Remove the comment, not accurate. But the code is good.

File: rust/src/enip/detect.rs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    return 0;
+}
+
+fn enip_cip_has_class(cipdir: &CipDir, class: u16) -> bool {
+    if let CipDir::Request(req) = cipdir {
+        for seg in req.path.iter() {
+            match seg.segment_type {
+                ENIP_CIP_PATH_CLASS_8BIT | ENIP_CIP_PATH_CLASS_16BIT => {
+                    if seg.value == class {
+                        return true;
```
Comment (jasonish) [inline rust/src/enip/detect.rs:201]: Can this conditional be put on the match arm? https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards?

File: rust/src/enip/enip.rs [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        }
+        _ => {
+            return false;
+        }
+    }
+}
+
+// C exports.
+
+unsafe extern "C" fn rs_enip_probing_parser_udp(
```
Comment (jasonish) [inline rust/src/enip/enip.rs:261]: Non-public, we can drop the `rs` prefix, just basic Rust naming.  Yeah, this is a break from the previous convention, which doesn't really make much sense anymore IMO.

File: rust/src/enip/mod.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+//! Application layer eniprs parser and logger module.
```
Comment (jasonish) [inline rust/src/enip/mod.rs:18]: typo: eniprs

Comment (catenacyber) [reply rust/src/enip/mod.rs:18]: Fixing this

File: rust/src/enip/enip.rs (hunk 3 above)
Comment (catenacyber) [reply rust/src/enip/enip.rs:261]: Ok for all cf `unsafe extern "C" fn  rs`

File: rust/src/enip/detect.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/enip/detect.rs:175]: ok

File: rust/src/enip/detect.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/enip/detect.rs:201]: Working locally, will try CI with MSRV next PR ;-)

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/9844

File: rust/src/enip/enip.rs (hunk 3 above)
Comment (jasonish) [reply rust/src/enip/enip.rs:261]: > Ok for all cf `unsafe extern "C" fn rs`

I think so.  Cause they don't become public linkable so don't form any part of the public C API.

**Pull request closed without merging**